### PR TITLE
Add Luau language

### PR DESCRIPTION
**Pull request by robloxiandemo**
## Description

Resolving [this issue](https://github.com/github-linguist/linguist/issues/6000#issue-1328038583), this PR aims to finally add Luau support to the platform.

## Checklist:

-   [x] **I am adding a new language.**
    -   [x] The extension of the new language is used in hundreds of repositories on GitHub.com.
        -   Search results for each extension:
            -   [.luau](https://github.com/search?q=path%3A*.luau&type=code&ref=advsearch) 3.9k files
    -   [x] I have included a real-world usage sample for all extensions added in this PR:
        -   Sample source(s):
            -   [EnumList.luau](https://github.com/Sleitnick/RbxUtil/blob/main/modules/enum-list/init.lua)
            -   [Option.luau](https://github.com/Sleitnick/RbxUtil/blob/main/modules/option/init.lua)
            -   [Ser.luau](https://github.com/Sleitnick/RbxUtil/blob/main/modules/ser/init.lua)
            -   [Symbol.luau](https://github.com/Sleitnick/RbxUtil/blob/main/modules/symbol/init.lua)
            -   [Tree.luau](https://github.com/Sleitnick/RbxUtil/blob/main/modules/tree/init.lua)
        -   Sample license(s):
            -   EnumList.luau - (https://github.com/Sleitnick/RbxUtil/blob/main/LICENSE.md) MIT
            -   Option.luau - (https://github.com/Sleitnick/RbxUtil/blob/main/LICENSE.md) MIT
            -   Ser.luau - (https://github.com/Sleitnick/RbxUtil/blob/main/LICENSE.md) MIT
            -   Symbol.luau - (https://github.com/Sleitnick/RbxUtil/blob/main/LICENSE.md) MIT
            -   Tree.luau - (https://github.com/Sleitnick/RbxUtil/blob/main/LICENSE.md) MIT
    -   [x] I have included a syntax highlighting grammar:
        -   https://github.com/JohnnyMorganz/Luau.tmLanguage – MIT
    -   [x] I have added a color:
        -   Hex value: `#14a5ff`
        -   Rationale: Logo color from https://luau-lang.org/ at the time of writing.
    -   [x] I have updated the heuristics to distinguish my language from others using the same extension.

### CODE REVIEW

**Comment by RobloxianDemo** [comment]
Hey! I'd like to ask if a CI could be ran for this, as I wasn't able to check it myself under the fork.

File: lib/linguist/languages.yml [hunk 1]
```diff
@@ -34,7 +34,6 @@
 # test changes in `test/test_blob.rb`.
 #
 # Please keep this list alphabetized. Capitalization comes before lowercase.
----
```
Comment (lildude) [inline lib/linguist/languages.yml:37]: Why are you removing this? Please don't.

File: lib/linguist/languages.yml [hunk 2]
```diff
@@ -3879,6 +3878,17 @@ Lua:
...
+  type: programming
+  tm_scope: source.luau
+  ace_mode: lua
+  codemirror_mode: lua
+  codemirror_mime_type: text/x-lua
+  color: "#14a5ff"
+  extensions:
+  - ".luau"
+  interpreters:
+  - lua
```
Comment (lildude) [inline lib/linguist/languages.yml:3891]: You've missed the step for adding the language ID. Also, this is used for scripts… is it common practice have shell scripts to be written in Luau?

**Review by lildude — CHANGES_REQUESTED** [review]

File: lib/linguist/languages.yml (hunk 2 above)
Comment (moo1210) [inline lib/linguist/languages.yml:3891]:
```suggestion
  - luau
```
Luau's binary is called luau, not lua.

**Review by moo1210 — CHANGES_REQUESTED** [review]

**Comment by moo1210** [comment]
I strongly support including support for Luau, [which for context is an open-source Lua 5.1 fork, with many backward-compatible features (and therefore syntax) changes built onto it over the years](https://luau-lang.org/).

With that being said, I do think it's worth noting that Vector3 (or any other Roblox types/globals) probably shouldn't be included in this. Standalone Luau has no built-in Vector3 type or globals such as game, this seems largely out of scope.

**Comment by RobloxianDemo** [comment]
I'll be going through to fix up my commits with your suggested changes, thank you!

File: lib/linguist/languages.yml (hunk 1 above)
Comment (RobloxianDemo) [reply lib/linguist/languages.yml:37]: Apologies for that, though it seems either my Prettier or another formatter extension was causing that on save, I'll add that back in my next few sets of commits.

File: lib/linguist/languages.yml (hunk 2 above)
Comment (RobloxianDemo) [reply lib/linguist/languages.yml:3891]: For whatever reason, none of the scripts were running on my system, and even after going through the hoops of installing Ruby. Would it be possible to determine a languageId manually?

As for the samples, I'll be working on adding more from public packages with real-world examples, apologies for not doing that sooner.

Comment (RobloxianDemo) [reply lib/linguist/languages.yml:3891]: Yeah, I wasn't sure what to put here, thank you for catching that!

**Comment by RobloxianDemo** [comment]
> I strongly support including support for Luau, [which for context is an open-source Lua 5.1 fork, with many backward-compatible features (and therefore syntax) changes built onto it over the years](https://luau-lang.org/).
> 
> With that being said, I do think it's worth noting that Vector3 (or any other Roblox types/globals) probably shouldn't be included in this. Standalone Luau has no built-in Vector3 type or globals such as game, this seems largely out of scope.

I will say that it seems the larger scope of usage of Luau is under the Roblox platform.

File: lib/linguist/languages.yml (hunk 2 above)
Comment (RobloxianDemo) [reply lib/linguist/languages.yml]: Resolved this issue [here](https://github.com/github-linguist/linguist/pull/6612/commits/428f5d21f1f95205d50e9e8f73aa009c4c078640), thank you for the contribution!

File: lib/linguist/languages.yml (hunk 1 above)
Comment (RobloxianDemo) [reply lib/linguist/languages.yml]: Resolved the codeblock issue [here](https://github.com/github-linguist/linguist/pull/6612/commits/428f5d21f1f95205d50e9e8f73aa009c4c078640).

File: lib/linguist/languages.yml (hunk 2 above)
Comment (lildude) [reply lib/linguist/languages.yml]: > For whatever reason, none of the scripts were running on my system, and even after going through the hoops of installing Ruby. 

You could or [preferred method and use a Codespace](https://github.com/github-linguist/linguist/blob/master/CONTRIBUTING.md#using-github-codespaces) - everyone gets some free usage.

> Would it be possible to determine a languageId manually?

You can run [this line](https://github.com/github-linguist/linguist/blob/7ca3799b8b5f1acde1dd7a8dfb7ae849d3dfb4cd/script/update-ids#L9) manually in irb to get the ID.

Comment (RobloxianDemo) [reply lib/linguist/languages.yml]: > > For whatever reason, none of the scripts were running on my system, and even after going through the hoops of installing Ruby.
> 
> You could use our [preferred method and use a Codespace](https://github.com/github-linguist/linguist/blob/master/CONTRIBUTING.md#using-github-codespaces) - everyone gets some free usage.

Alright, I'll give a shot with Codespace, thank you!

> > Would it be possible to determine a languageId manually?
> 
> You can run [this line](https://github.com/github-linguist/linguist/blob/7ca3799b8b5f1acde1dd7a8dfb7ae849d3dfb4cd/script/update-ids#L9) manually in irb to get the ID.

IRB doesn't want to run that, so I'll stick to trying on a Codespace.

Comment (RobloxianDemo) [reply lib/linguist/languages.yml]: Resolved [this languageId issue here](https://github.com/github-linguist/linguist/pull/6612/commits/ba342d6f93708df835cd9e78f7f91d6e118e71cb).

**Comment by RobloxianDemo** [comment]
> > ```
> >     * Hex value: `#14a5ff`
> >     * Rationale: Logo color from https://luau-lang.org/ at the time of writing.
> > ```
> 
> Any chance you could update the color to #479df9 ? Not sure where you got yours from image image

Sure, I'll submit a commit for this now.

**Comment by lewisakura** [comment]
I do hate to add to the color debate, but neither of those colors is correct for the logo. The logo uses #00A2FF:

https://github.com/luau-lang/site/blob/c4e26ca38e3d075a365f548951f05a9be0eda345/logo.svg#L2

**Comment by RobloxianDemo** [comment]
> I do hate to add to the color debate, but neither of those colors is correct for the logo. The logo uses #00A2FF:
> 
> https://github.com/luau-lang/site/blob/c4e26ca38e3d075a365f548951f05a9be0eda345/logo.svg?short_path=5525394#L2

Haha, thank you for pointing that out, I'll write up a commit later today to resolve that. 👍

**Comment by RobloxianDemo** [comment]
> I do hate to add to the color debate, but neither of those colors is correct for the logo. The logo uses #00A2FF:
> 
> https://github.com/luau-lang/site/blob/c4e26ca38e3d075a365f548951f05a9be0eda345/logo.svg?short_path=5525394#L2

[Resolved here](https://github.com/github-linguist/linguist/pull/6612/commits/92bed6ff79b75156515a7d3d2328a7fb17403495), thank you for your contribution!

**Comment by RobloxianDemo** [comment]
Seems we'll have to make some changes with the grammar. I'll try to submit a few PRs later this week to resolve it.

**Comment by lildude** [comment]
> Seems we'll have to make some changes with the grammar. I'll try to submit a few PRs later this week to resolve it.

Really? The current test failures aren't due to the grammar… they're ordering issues caused by doing things manually early on in the PR.

**Comment by RobloxianDemo** [comment]
> > Seems we'll have to make some changes with the grammar. I'll try to submit a few PRs later this week to resolve it.
> 
> Really? The current test failures aren't due to the grammar… they're ordering issues.

I ran some private tests with the grammar shortly after seeing those failures, and eleven errors had shown up, apparently due to some backtracking issues with the language. I'd have to check out what's causing it, though I'm not entirely sure.

File: samples/Luau/none.luau [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+--[=[
+	@prop None None
+	@within Freeze
+
+	Since lua tables cannot distinguish between values not being present and a value of nil,
+	`Freeze.None` exists to represent values that should be interpreted as `nil`.
+
+	This is useful when removing values with functions such as [`Freeze.Dictionary.merge`](../api/Dictionary#merge).
+]=]
```
Comment (eric-wieser) [inline samples/Luau/none.luau:13]: @lildude: based on your comment on #6616, should the tokenizer be taught about `--[[ ]]` and `--[====[`/`]====]` comments? Or is that only necessary if there are extension ambiguities?

Comment (lildude) [reply samples/Luau/none.luau:13]: There's probably no harm in adding support for this format but it only really comes into play when there are extension ambiguities.

**Comment by lildude** [comment]
> I ran some private tests with the grammar shortly after seeing those failures, and eleven errors had shown up, apparently due to some backtracking issues with the language. I'd have to check out what's causing it, though I'm not entirely sure.

I see it now. It's because of the `new/NewLuau.YAML-tmLanguage` file. Linguist isn't fussy about the location of the `*.YAML-tmLanguage` files hence it is finding it and attempting to convert it. It doesn't look like this file is intended to be used yet so the best thing would be for the grammar maintainer to move this file into it's own branch so we can use the main branch without it causing troubles.

**Comment by JohnnyMorganz** [comment]
Please open a PR for it! I would be happy to accept. I don't have suitable access to do it myself right now.

Although when I last tried to add to linguist locally I also experienced issues where the linguist TextMate grammar parser does not support non-fixed lookbehind (whilst VSCode does). Would be awesome to see if it works though! (Removing the non-fixed lookbehind requires non trivial amounts of effort so I did not get round to it)

**Comment by moo1210** [comment]
This seems more like a parsing issue, which happens regardless of which yaml file exists.
```
- [ ] repository `vendor/grammars/Luau.tmLanguage` (from https://github.com/moo1210/Luau.tmLanguage.git) (10 errors)
  - Invalid regex in grammar: `source.luau` (in `Luau.YAML-tmLanguage`) contains a malformed regex (regex "`(?<![:>.]\s*)\b_\b`": lookbehind assertion is not fixed length (at offset 12))
  - Invalid regex in grammar: `source.luau` (in `Luau.YAML-tmLanguage`) contains a malformed regex (regex "`(?<![:>.]\s*)\b([A-Z_][A-Z0-9_]*`...": lookbehind assertion is not fixed length (at offset 12))
  - Invalid regex in grammar: `source.luau` (in `Luau.YAML-tmLanguage`) contains a malformed regex (regex "`(?<![:>.]\s*)\b([a-zA-Z_]\w*)\b`": lookbehind assertion is not fixed length (at offset 12))
  - Invalid regex in grammar: `source.luau` (in `Luau.YAML-tmLanguage`) contains a malformed regex (regex "`(?<!:\s*|->\s*)\b([a-zA-Z_]\w*)\`...": lookbehind assertion is not fixed length (at offset 8))
  - Invalid regex in grammar: `source.luau` (in `Luau.YAML-tmLanguage`) contains a malformed regex (regex "`(?<![:>.]\s*)\b_\b`": lookbehind assertion is not fixed length (at offset 12))
  - Invalid regex in grammar: `source.luau` (in `Luau.YAML-tmLanguage`) contains a malformed regex (regex "`(?<![:>.]\s*)\b([A-Z_][A-Z0-9_]*`...": lookbehind assertion is not fixed length (at offset 12))
  - Invalid regex in grammar: `source.luau` (in `Luau.YAML-tmLanguage`) contains a malformed regex (regex "`(?<![:>.]\s*)\b([a-zA-Z_]\w*)\b`": lookbehind assertion is not fixed length (at offset 12))
  - Invalid regex in grammar: `source.luau` (in `Luau.YAML-tmLanguage`) contains a malformed regex (regex "`(?<![:>.]\s*)\b_\b`": lookbehind assertion is not fixed length (at offset 12))
  - Invalid regex in grammar: `source.luau` (in `Luau.YAML-tmLanguage`) contains a malformed regex (regex "`(?<![:>.]\s*)\b([A-Z_][A-Z0-9_]*`...": lookbehind assertion is not fixed length (at offset 12))
  - Invalid regex in grammar: `source.luau` (in `Luau.YAML-tmLanguage`) contains a malformed regex (regex "`(?<![:>.]\s*)\b([a-zA-Z_]\w*)\b`": lookbehind assertion is not fixed length (at offset 12))
  ```

**Comment by lildude** [comment]
> Although when I last tried to add to linguist locally I also experienced issues where the linguist TextMate grammar parser does not support non-fixed lookbehind (whilst VSCode does)

If it helps, this is because the highlighting engine uses PCRE for performance reasons and as such the compiler enforces the PCRE rules.

**Comment by JohnnyMorganz** [comment]
Yup that's fair, same errors that I see.

The grammar in its current state is unsuitable. Fixing the grammar will require some considerable changes to make it compatible (it may even be simpler to just rewrite it)

**Comment by AsynchronousAI** [comment]
I believe Luau should be an option for better syntax highlighting supporting types, format strings, etc. and seperating Roblox projects from normal Lua projects in trending or search.

**Comment by CompeyDev** [comment]
> The grammar in its current state is unsuitable. Fixing the grammar will require some considerable changes to make it compatible (it may even be simpler to just rewrite it)

Any updates on this? Would be SUPER cool to have luau support in discovery and syntax highlighting.

**Comment by JohnnyMorganz** [comment]
> Any updates on this?

Nope, unfortunately I don't have the time to work on that grammar right now. Contributions welcome!

**Comment by JohnnyMorganz** [comment]
I've fixed the grammar so it should now support Linguist. Can you pull latest changes @robloxiandemo?

Also, I expect to still make improvements. I think I already read this somewhere, but just to double check, grammars automatically get updated in the repo (before the next version release)?

**Comment by Kikki005** [comment]
Help Get it back 0xb6270e842c3Fc3A2f55683C43C16A83fFe27a6Be

**Review by AsynchronousAI — APPROVED** [review]
LGTM, I also checked the TextMate grammar!  (I guess my opinion doesn't really matter cause im not a maintainer but anyways)

**Comment by robloxiandemo** [comment]
> I've fixed the grammar so it should now support Linguist. Can you pull latest changes @robloxiandemo?

Hey, and good stuff! Although, I have been trying to pull them, though it just refuses to update to the following regex-related error:
image

I'm unsure on how or why it's attempting to use the nonexistent ["NewLuau.YAML-tmLanguage"](https://github.com/JohnnyMorganz/Luau.tmLanguage/commit/bc4170802aab182127f2b6b80b0e1164f50d78c1) file. Any ideas on how I could get around this, or simply to resolve it?

**Comment by JohnnyMorganz** [comment]
This file was already deleted. Can you double check the submodule was updated appropriately? What git revision is checked out

**Comment by robloxiandemo** [comment]
> This file was already deleted. Can you double check the submodule was updated appropriately? What git revision is checked out

I checked and it seems to be pulling [this](https://github.com/JohnnyMorganz/Luau.tmLanguage/commit/9f320add6dd099f99897a03c5e2ce3ca5f176d7f) as the latest revision, a bit stumped on how to get around this to retrieve the repo's latest commit.
Screenshot 2024-04-03 063236

**Comment by checkraisefold** [comment]
Great stuff that this is finally being completed!

**Review by AsynchronousAI — APPROVED** [review]

**Comment by lildude** [comment]
Not until you resolve the conflict 😉

**Comment by robloxiandemo** [comment]
So whenever I try to add this to the vendor/grammar list of languages, it seems to introduce a merge conflict, so I'm unsure on where to proceed.

**Comment by lildude** [comment]
Git should prompt you to resolve the conflicts if you merge master into your branch. Alternatively use the GitHub "Resolve Conflicts" button and do it in your browser.

**Comment by robloxiandemo** [comment]
> Git should prompt you to resolve the conflicts if you merge master into your branch. Alternatively use the GitHub "Resolve Conflicts" button and do it in your browser.

I'll try that now, thank you.

**Comment by robloxiandemo** [comment]
Alright, it seems to have resolved now, that's real interesting.

File: vendor/licenses/git_submodule/Luau.tmLanguage.dep.yml [hunk 4]
```diff
@@ -0,0 +1,57 @@
+---
+name: Luau.tmLanguage
+version: bc4170802aab182127f2b6b80b0e1164f50d78c1
+type: git_submodule
+homepage: https://github.com/JohnnyMorganz/Luau.tmLanguage.git
+license: other
```
Comment (lildude) [inline vendor/licenses/git_submodule/Luau.tmLanguage.dep.yml:6]: You can manually change this as licensed struggles a bit when it encounters slight variations in the licence files.

```suggestion
license: mit
```

**Review by lildude — CHANGES_REQUESTED** [review]
We really don't need that many samples; two or three showing different elements of the language is sufficient. Additionally we only want real world samples. I don't know the language, but several of your samples appear to be contrived illustrations on how to use the language, rather than actual "used in the wild" code.

**Comment by robloxiandemo** [comment]
> We really don't need that many samples; two or three showing different elements of the language is sufficient. Additionally we only want real world samples. I don't know the language, but several of your samples appear to be contrived illustrations on how to use the language, rather than actual "used in the wild" code.

I understand your suggestions and have updated the related samples' source to provide more up-to-date and more demonstrative examples on how the language is used in the real world.

File: vendor/licenses/git_submodule/Luau.tmLanguage.dep.yml (hunk 4 above)
Comment (robloxiandemo) [reply vendor/licenses/git_submodule/Luau.tmLanguage.dep.yml]: > You can manually change this as licensed struggles a bit when it encounters slight variations in the license files.

I hadn't caught onto that earlier, though it's helpful, thank you! Is there anything in particular that prevents it from detecting the relevant license file?

**Review by lildude — CHANGES_REQUESTED** [review]
Your new samples are all now too big. If they're suppressed in the diff, they're bigger than we need. Also, we don't need all the other grammar updates.

**Comment by checkraisefold** [comment]
looks like you updated every submodule from remote instead of just the Luau grammar

**Comment by AsynchronousAI** [comment]
> We really don't need that many samples; two or three showing different elements of the language is sufficient. Additionally we only want real world samples. I don't know the language, but several of your samples appear to be contrived illustrations on how to use the language, rather than actual "used in the wild" code.

Only one of these are "used in the wild" and that is Promise. I think the rest should be removed and maybe adding @sleitnick's [RbxUtil](https://github.com/Sleitnick/RbxUtil) should be a better choice when looking for examples since they are small and used **very often**

**Comment by robloxiandemo** [comment]
> Your new samples are all now too big. If they're suppressed in the diff, they're bigger than we need. Also, we don't need all the other grammar updates.

It seems I had overlooked my changes, will start reverting and correcting the issues now, thank you.

**Comment by robloxiandemo** [comment]
> looks like you updated every submodule from remote instead of just the Luau grammar

That was a mistake, I had thought that it was cleared from my git commit tree upon fetching, though it seems it hadn't been discarded properly.

**Comment by lildude** [comment]
> > Your new samples are all now too big. If they're suppressed in the diff, they're bigger than we need. Also, we don't need all the other grammar updates.
> 
> Is there anything else that still needs changing? 🤔

Yes. All the submodule updates need to be reverted. We only need the files for this language updated.

**Comment by AsynchronousAI** [comment]
> > > Your new samples are all now too big. If they're suppressed in the diff, they're bigger than we need. Also, we don't need all the other grammar updates.
> > 
> > 
> > Is there anything else that still needs changing? 🤔
> 
> Yes. All the submodule updates need to be reverted. We only need the files for this language updated.

Ight bet, Im on a codespace rn fixing.

**Comment by CompeyDev** [comment]
@robloxiandemo If you're looking for more samples, I have a project, [rusty-luau](https://github.com/CompeyDev/rusty-luau), which uses a lot of luau-specific features (such as types), consider that!

**Comment by deviaze** [comment]
Most of those examples still seem to showcase relatively advanced type features that I don't think represent the bulk of the language,

Luau is an extremely useful scripting language, not just on the Roblox platform, but also for general purpose programming with the Lune runtime. It's an amazing replacement for bash/powershell scripts, python, and other high level dynamically typed scripting languages--without needing any of the baggage to install or run. Imho, Luau excels in its simplicity, ease of use, and performance.

If possible, I'd like to contribute some simple, general-purpose script examples that demonstrate this side of the language but still differentiate it from Lua 5.1.

**Comment by robloxiandemo** [comment]
> @robloxiandemo If you're looking for more samples, I have a project, [rusty-luau](https://github.com/CompeyDev/rusty-luau), which uses a lot of luau-specific features (such as types), consider that!

I'll look over them and update each, though I've done it so many times already, 😅

**Comment by robloxiandemo** [comment]
> If possible, I'd like to contribute some simple, general-purpose script examples that demonstrate this side of the language but still differentiate it from Lua 5.1.

Feel free to, just make sure to mention me somewhere along with the example and any related real-world usage, been finding it a tad bit difficult to find outside of Roblox examples without there just being comment spam (or having far too many lines).

**Review by lildude — APPROVED** [review]
LGTM. Thanks.

Note: this PR will not be merged until close to when the next release is made. See [here](https://github.com/github/linguist/blob/master/docs/troubleshooting.md#my-linguist-pr-has-been-merged-but-github-doesnt-reflect-my-changes) for more details.

**Comment by bradsharp** [comment]
Color is wrong, it should be #00A2FF

Source: I produced the Luau logo

**Pull request merged**